sno/checkout.py

```diff
@@ -307,6 +307,41 @@ def restore(ctx, source, pathspec):
     )
 
 
+@click.command()
+@click.pass_context
+@click.option(
+    "--discard-changes",
+    "--force",
+    "-f",
+    is_flag=True,
+    help="Discard local changes in working copy if necessary",
+)
+@click.argument("refish")
+def reset(ctx, discard_changes, refish):
+    """
+    Reset the branch head to point to a different commit.
+    """
+    repo = ctx.obj.repo
+
+    try:
+        commit_or_tree, ref = repo.resolve_refish(refish)
+        commit = commit_or_tree.peel(pygit2.Commit)
+    except (KeyError, pygit2.InvalidSpecError):
+        raise NotFound(f"{refish} is not a commit", exit_code=NO_COMMIT)
+
+    same_commit = repo.head_commit == commit
+    if not discard_changes and not same_commit:
```

Comment:
`and not same_commit` - doesn't that mean that we're not properly checking the repo is not dirty if we happen to already be on the target commit? I think this clause could be removed

Reply:
Your understanding of the logic seems to be correct. This behaviour is consistent with other commands, eg:

Sno checkout:

$ sno checkout HEAD^
Error: You have uncommitted changes in your working copy.
Commit these changes first (`sno commit`) or just discard them by adding the option `--discard-changes`.
$ sno checkout HEAD
(no problem)
$ sno checkout -b new_branch   # kind of like checkout HEAD, but also actually does something
(no problem)

Sno reset:

$ sno reset HEAD^ 
Error: You have uncommitted changes in your working copy.
Commit these changes first (`sno commit`) or just discard them by adding the option `--discard-changes`.
$ sno reset HEAD
(no problem)

Of course, `sno reset HEAD` is a no-op. But that's no reason to make it not work. But in the case of sno reset, there's only two options - either we need to discard-changes, or the command is a no-op (unlike sno checkout above, which can still have useful side effects). So, maybe we shouldn't do this check. But the check is consistent with other commands... what do you think?

Reply:
I think it's weird, but okay. It's strange because git actually _does_ allow you to checkout and reset all you like while you have local changes, as long as those changes don't conflict with the tip change. Is the reason that we don't allow that, something to do with what you alluded to above about schema changes? Or was it just difficult to reason about so we never bothered to implement it so far?

Reply:
Moved discussion offline